pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,239 @@
+		if err != nil {
+			panic(err)
+		}
+		return ecdsaKey.Curve == *curve
+	case ED25519:
+		_, ok := pubKey.(ed25519.PublicKey)
+		return ok
+	}
+	panic("unreachable")
+}
+
+func (a algorithmDetailsImpl) checkHash(hashType crypto.Hash) bool {
+	return a.hashType == hashType
+}
+
+var algorithmDetails = []algorithmDetailsImpl{
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_2048_SHA256, RSA, crypto.SHA256, RSAKeySize(2048), "rsa-sign-pkcs1-2048-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_3072_SHA256, RSA, crypto.SHA256, RSAKeySize(3072), "rsa-sign-pkcs1-3072-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_4096_SHA256, RSA, crypto.SHA256, RSAKeySize(4096), "rsa-sign-pkcs1-4096-sha256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_256_NISTP256, ECDSA, crypto.SHA256, elliptic.P256(), "ecdsa-sha2-256-nistp256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_384_NISTP384, ECDSA, crypto.SHA384, elliptic.P384(), "ecdsa-sha2-384-nistp384"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_512_NISTP521, ECDSA, crypto.SHA512, elliptic.P521(), "ecdsa-sha2-512-nistp521"},
+	{v1.KnownSignatureAlgorithm_ED25519, ED25519, crypto.Hash(0), nil, "ed25519"},
+	{v1.KnownSignatureAlgorithm_ED25519_PH, ED25519, crypto.SHA512, nil, "ed25519-ph"},
+}
+
+// AlgorithmRegistryConfig represents a set of permitted algorithms for a given Sigstore service or component.
+//
+// Individual services may wish to restrict what algorithms are allowed to a subset of what is covered in the algorithm
+// registry (represented by v1.KnownSignatureAlgorithm).
+type AlgorithmRegistryConfig struct {
+	permittedAlgorithms []AlgorithmDetails
+}
+
+// GetAlgorithmDetails retrieves a set of details for a given v1.KnownSignatureAlgorithm flag that allows users to
+// introspect the public key algorithm, hash algorithm and more.
+func GetAlgorithmDetails(knownSignatureAlgorithm v1.KnownSignatureAlgorithm) (AlgorithmDetails, error) {
+	for _, detail := range algorithmDetails {
+		if detail.knownAlgorithm == knownSignatureAlgorithm {
+			return &detail, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find algorithm details for known signature algorithm: %s", knownSignatureAlgorithm)
+}
+
+// NewAlgorithmRegistryConfig creates a new AlgorithmRegistryConfig for a set of permitted signature algorithms.
+func NewAlgorithmRegistryConfig(algorithmConfig []v1.KnownSignatureAlgorithm) (*AlgorithmRegistryConfig, error) {
+	permittedAlgorithms := make([]AlgorithmDetails, 0, len(algorithmDetails))
+	for _, algorithm := range algorithmConfig {
+		a, err := GetAlgorithmDetails(algorithm)
+		if err != nil {
+			return nil, err
+		}
+		permittedAlgorithms = append(permittedAlgorithms, a)
+	}
+	return &AlgorithmRegistryConfig{permittedAlgorithms: permittedAlgorithms}, nil
+}
+
+// IsAlgorithmPermitted checks whether a given public key/hash algorithm combination is permitted by a registry config.
+func (registryConfig AlgorithmRegistryConfig) IsAlgorithmPermitted(key crypto.PublicKey, hash crypto.Hash) bool {
```

Comment:
How do you foresee this being used with KMS keys? Should the caller be required to inspect the KMS key for its key and hash (we have key, we're missing hash, but it's being implemented in https://github.com/sigstore/sigstore/pull/1426). Should each KMS provider has its own registry?(as a service, this would be a bit annoying to use, since you'd have to consult a registry of regstries)

Reply:
I think it makes most sense for the calling code to inspect the KMS key and hash, and check against the registry to see whether it should use that key or not.

I don't think it makes sense for each KMS provider to own its own registry, but perhaps the `kms.SignerVerifier` API could take an optional `AlgorithmRegistryConfig` and do the checking there? But I'd lean towards not tangling the APIs together like that and just let the calling code compose these calls as required.

What are your thoughts?

Reply:
That sounds good to me. Long-term I'd lean towards the proposal to take in a config, just to avoid each client having to reimplement this, but I don't think it's something that's blocking regardless, just thinking through how we'll handle users who provide their own KMS keys.